taichi/ir/type.h

```diff
@@ -552,6 +631,99 @@ class TypedConstant {
   float64 val_cast_to_float64() const;
 };
 
+template <typename S, typename T>
+typename std::enable_if<std::is_base_of_v<Type, T>, void>::type
+Type::ptr_io(const T *&ptr, S &serializer, bool writing) {
+  if (writing) {
+    if (ptr == nullptr) {
+      serializer("type_kind", (TypeKind)-1);
+      return;
+    }
+    serializer("type_kind", ptr->type_kind);
+    if (false) {
```

Comment:
debug code?

Reply:
there is sth alike in the json part, you might want to remove that too

Reply:
this `if false` is for the `else if` in the macro below to work

Reply:
nit: Add a comment here to say this `if (false)` is for the `PER_TYPE_KIND` macro.

Although... why isn't this macro a switch case statement?